OpenRA.Game/Network/OrderManager.cs

```diff
@@ -181,15 +178,41 @@ public IEnumerable<Session.Client> GetClientsNotReadyForNextFrame
 			}
 		}
 
-		public void Tick()
+		void SendOrders()
 		{
-			if (!IsReadyForNextFrame)
-				throw new InvalidOperationException();
+			if (NetFrameNumber < 1)
+				return;
 
-			if (GameSaveLastFrame < NetFrameNumber + FramesAhead)
-				Connection.Send(NetFrameNumber + FramesAhead, localOrders.Select(o => o.Serialize()).ToList());
+			// Loop exists to ensure we never miss a frame, since that would stall the game.
+			// This loop also sends the initial blank frames to get to the correct order latency.
+			while (lastFrameSent < NetFrameNumber + FramesAhead)
```

Comment:
Is this loop doing anything during normal gameplay, or is this just to replace the loop in the ctor?

Reply:
it looks like this may also be fufilling the role of `isReadyForNextFrame`? It would IMO be clearer if we could keep the loop sending empty orders in the ctor, then inline the `if (...) Send() Clear()` inside the `if (!isReadyForNextFrame)` in `SendNetFrameOrdersAndCheckReady`.

What do you think?

Reply:
I've changed the code a bit since, but the idea behind having the loop with the send is that all the logic that ensures the correct number of orders are sent for our given configuration (`FramesAhead` and `CurrentNetFrame`) is kept together.

In the adaptive order latency approach, FramesAhead would change, and in the server-sync approach, it wouldn't exist anymore, so I'd like to keep it more or less hidden from the general structure of "frame readiness". This approach cuts it down at least.

Happy to change if you disagree though.